Adds documentation for mod1 in mathematical operations. Fixes #15486

mronian opened the pull request:
Also there is a formatting issue for mod http://docs.julialang.org/en/release-0.4/stdlib/math/#Base.mod where the '``' are visible. Corrected that.

eschnett commented on base/docs/helpdb/Base.jl:
```diff
@@ -2265,11 +2265,11 @@ sortperm
     mod2pi(x)
 
-Modulus after division by 2pi, returning in the range \[0,2pi).
+Modulus after division by `2pi`, returning in the range `\[0,2pi)`.
```
These changes seem unrelated?

nalimilan replied:
...but they are in a separate commit.

mronian replied:
Should I make a separate PR?

eschnett replied:
I'd be fine with this, but I've encountered stiff resistance in the past to adding unrelated changes to a PR.

nalimilan replied:
For small uncontroversial PRs, I would think it's fine. Else it will increase the cost of fixing small mistakes you find while reading the docs, and nobody will do it.

tkelman replied:
these should probably use latex double backticks, since the second case is mathematical notation not julia source

mronian replied:
Do you mean something like this ?

MichaelHatherly replied:
Use backticks both before and after the `2pi`, no `''`s:

```
Modulus after division by ``2π``, returning in the range ``[0, 2π)``.
```

Might as well use a unicode pi as well, and the `\` isn't needed.

eschnett commented on doc/manual/mathematical-operations.rst:
```diff
@@ -454,6 +454,7 @@ Function                     Description
 :func:`cld(x,y) <cld>`       ceiling division; quotient rounded towards ``+Inf``
 :func:`rem(x,y) <rem>`       remainder; satisfies ``x == div(x,y)*y + rem(x,y)``; sign matches ``x``
 :func:`mod(x,y) <mod>`       modulus; satisfies ``x == fld(x,y)*y + mod(x,y)``; sign matches ``y``
+:func:`mod1(x,y) <mod1>`     modulus with offset 1; gives a result ``r`` such that ``mod(r, y)==mod(x,y)``; sign matches ``y``
```
The documentation of `mod1` should use the function name `mod1` as well.

`mod(r, y) == mod(x,y)` is a post-condition, but is not sufficient to unambiguously define the result. I believe there is a good definition elsewhere. Here is what the source code says:
```Julia
# mod returns in [0,y) or (y,0] (for negative y),
# whereas mod1 returns in (0,y] or [y,0)
```

mronian replied:
I found this definition in Mathematica's docs. Very few other programming languages seem to have mod with offset.

eschnett replied:
It's not quite a definition since it doesn't define `r`; the condition you quote (which is quite elegant, I didn't realize this before) allows for many values of `r`. If you combine it with the condition `0 < r <= y` (for positive `y`) etc. you'll have a complete definition. You might even want to update the existing documentation of `mod1` with this.

mronian replied:
I think it works for negative ```y``` as well.

mronian replied:
Is this fine?
```
Modulus after flooring division, returning a value `r` such that `mod(r, y)==mod(x,y)` in the range `(0, y]`
```

eschnett replied:
For negative `y`, the result is in the rage `[y, 0)`, which is different.

eschnett commented on doc/stdlib/math.rst:
```diff
@@ -134,7 +134,7 @@ Mathematical Operators
    .. Docstring generated from Julia source
 
-   Modulus after flooring division, returning in the range [0,``y``\ ), if ``y`` is positive, or (``y``\ ,0] if ``y`` is negative.
+   Modulus after flooring division, returning in the range ``[0,y)``\ , if ``y`` is positive, or ``(y,0]`` if ``y`` is negative.
```
Isn't the `rst` documentation auto-generated?

mronian replied:
Yes. It was generated after re-building julia.

mronian commented:
@eschnett I've amended the last commit. Please review.

mronian commented:
@MichaelHatherly I've made the changes and it looks much better now. Thanks! 

I've amended the previous commit. Please review the PR.

MichaelHatherly commented on base/docs/helpdb/Base.jl:
```diff
@@ -2265,11 +2265,11 @@ sortperm
     mod2pi(x)
 
-Modulus after division by 2pi, returning in the range \[0,2pi).
+Modulus after division by ``2π``, returning in the range ``\[0,2π)``.
```
The `\` char isn't needed, sphinx seems to manage find without them. Also the ones further down, `\[0, y)` etc.

MichaelHatherly commented on base/docs/helpdb/Base.jl:
```diff
@@ -2265,11 +2265,11 @@ sortperm
-numerically exact 2pi, and is therefore not exactly the same as mod(x,2pi), which would
-compute the modulus of `x` relative to division by the floating-point number 2pi.
+numerically exact ``2π``, and is therefore not exactly the same as `mod(x,2π)`, which would
+compute the modulus of `x` relative to division by the floating-point number ``2π``.
```
Since the `2π` at the end of this doc is referring to the floating point value rather than mathematical 2π perhaps it should just be single backticks:

```
... floating-point number `2π`.
```

MichaelHatherly commented on base/docs/helpdb/Base.jl:
```diff
@@ -8089,7 +8089,7 @@ fld1
     mod1(x, y)
 
-Modulus after flooring division, returning a value in the range `(0, y]`.
+Modulus after flooring division, returning a value `r` such that `mod(r, y)==mod(x,y)` in the range `(0, y]` for positive `y` and in the range `[y,0)` for negative `y`.
```
Docstrings should be hard-wrapped at 92 chars if possible.

tkelman replied:
still needs wrapping, and should be consistent - if `mod(r, y)` has a space after the comma then `mod(x,y)` should too (or neither), and the ranges should be math quoted here too

mronian commented:
@MichaelHatherly I've made the changes. Please review the PR.

tkelman commented on base/docs/helpdb/Base.jl:
```diff
@@ -2265,11 +2265,11 @@ sortperm
     mod2pi(x)
 
-Modulus after division by 2pi, returning in the range \[0,2pi).
+Modulus after division by ``2π``, returning in the range ``[0,2π)``.
```
should pick and be consistent whether these should be highlighted as code (single backticks in markdown) or as latex-style math (double backticks in our markdown flavor)

mronian replied:
@tkelman Which one should I use?

tkelman replied:
latex math for the ranges since that notation isn't julia code, code for the rest

mronian committed:
Adds documentation for mod1 and corrects formatting for mod2pi, mod.
Shifts documentation of  mod1, fld, fldmod1, mod2pi from helpdb/Base.jl to
the respective function definition files.

tkelman commented on base/docs/helpdb/Base.jl:
```diff
@@ -7601,8 +7601,8 @@ procs(::SharedArray)
-Modulus after flooring division, returning in the range \[0,`y`), if `y` is positive, or
-(`y`,0\] if `y` is negative.
+Modulus after flooring division, returning in the range `[0,y)`, if `y` is positive, or
+`(y,0]` if `y` is negative.
```
same applies to these ranges, this isn't julia syntax so use latex math style double backticks

MichaelHatherly commented:
Sorry for missing that last ping @mronian. Other than Tony's comments, would be good to move these docs out of `helpdb/Base.jl` and put them with their actual function definitions.

tkelman commented on doc/manual/mathematical-operations.rst:
```diff
@@ -463,6 +463,7 @@ Function                     Description
 :func:`cld(x,y) <cld>`       ceiling division; quotient rounded towards ``+Inf``
 :func:`rem(x,y) <rem>`       remainder; satisfies ``x == div(x,y)*y + rem(x,y)``; sign matches ``x``
 :func:`mod(x,y) <mod>`       modulus; satisfies ``x == fld(x,y)*y + mod(x,y)``; sign matches ``y``
+:func:`mod1(x,y) <mod1>`     ``mod()`` with offset 1; returns ``r∈(0,y]`` for ``y>0`` or ``r∈[y,0)`` for ``y<0`` & ``mod(r, y)=mod(x,y)``
```
the meaning of the `&` isn't obvious here, maybe replace it with ```... or ``r∈[y,0)`` for ``y<0``, where ``mod(r,y)==mod(x,y)`` ```

tkelman commented on base/int.jl:
````diff
@@ -88,9 +88,28 @@ rem(x::Unsigned, y::Signed) = rem(x,unsigned(abs(y)))
+```julia
+x == fld(x,y)*y + mod(x,y)
+```
+"""
````
the docstring should be immediately preceding the most general method, without an empty line between the end of the docstring and the start of the method definition

tkelman replied:
the most general method is the `T<:Integer` one

tkelman commented on doc/manual/mathematical-operations.rst:
```diff
@@ -463,6 +463,7 @@ Function                     Description
 :func:`cld(x,y) <cld>`       ceiling division; quotient rounded towards ``+Inf``
 :func:`rem(x,y) <rem>`       remainder; satisfies ``x == div(x,y)*y + rem(x,y)``; sign matches ``x``
 :func:`mod(x,y) <mod>`       modulus; satisfies ``x == fld(x,y)*y + mod(x,y)``; sign matches ``y``
+:func:`mod1(x,y) <mod1>`     ``mod()`` with offset 1; returns ``r∈(0,y]`` for ``y>0`` or ``r∈[y,0)`` for ``y<0``, where ``mod(r, y)=mod(x,y)``
```
`mod(r, y) == mod(x, y)`

mronian commented:
@tkelman Please Review. I g

mronian commented:
@tkelman Why did the AppVeyor build fail? I could not understand from the logs.

tkelman commented:
Not sure. Looks like some processes ran out of memory and crashed in an unclean way https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.2255/job/f9xlgf9uepi8wmxn

Almost certainly not your fault, but restarted just in case. This looks good now.

mronian commented:
The build has failed again for some reason.

zhmz90 commented:
How to restart the build? I didn't find it.

mronian commented:
@tkelman Its passing now. Please review.
